acceptance/suites/tests/00_smoke/puppetdb_integration.rb

```diff
@@ -107,13 +107,38 @@
   end
 end
 
-step 'Validate server sent agent data to PuppetDB' do
+step 'Validate PuppetDB successfully stored agent data' do
   query = "curl http://localhost:8080/pdb/query/v4/nodes/#{master_fqdn}"
-  response = JSON.parse(on(master, query).stdout.chomp)
-  %w[facts_timestamp catalog_timestamp report_timestamp].each do |dataset|
-    assert_operator(Time.iso8601(response[dataset]),
-                    :>,
-                    run_timestamp,
-                   "#{dataset} updated in PuppetDB")
+  agent_datasets = %w[facts_timestamp catalog_timestamp report_timestamp]
+  missing_datasets = [ ]
+  retries = 3
+
+  retries.times do |i|
+    logger.debug("PuppetDB query attempt #{i} for updated agent data...")
+
+    missing_datasets = [ ]
+    response = JSON.parse(on(master, query).stdout.chomp)
+
+    dataset_states = response.select {|k, v| agent_datasets.include?(k)}.map do |k, v|
+      t = Time.iso8601(v) rescue nil
+      [k, t]
+    end.to_h
+
+    missing_datasets = dataset_states.select {|k, v| v.nil? || (v < run_timestamp)}.keys
```

Comment:
Should we be asserting that the value is greaterthan the run_timestamp here?

Reply:
Nope. This is checking for datasets that didn't make it in: either they're `null` because no data is there, or the timestamp is older than the time taken before the puppet run.

Later on, we assert that the result of this selection is empty.